src/interface/timer.rs

```diff
@@ -26,3 +29,104 @@ pub fn readtimer(now: RustInstant) -> RustDuration {
 pub fn sleep(dur: RustDuration) {
     thread::sleep(dur);
 }
+
+#[derive(Debug)]
+struct _IntervalTimer {
+    pub cageid: u64,
+    pub init_instant: RustInstant, // The instant this process is created
+    
+    pub start_instant: RustInstant,
+    pub curr_duration: RustDuration,
+    pub next_duration: RustDuration,
+
+    pub is_ticking: bool,
+}
+
+#[derive(Clone, Debug)]
+pub struct IntervalTimer {
+    _ac: Arc<Mutex<_IntervalTimer>>,
+}
+
+impl IntervalTimer {
+    pub fn new(cageid: u64) -> Self {
+        Self {
+            _ac: Arc::new(Mutex::new(
+                _IntervalTimer {
+                    cageid: cageid,
+                    init_instant: RustInstant::now(),
+                    start_instant: RustInstant::now(),
+                    curr_duration: RustDuration::ZERO,
+                    next_duration: RustDuration::ZERO,
+                    is_ticking: false,
+                }
+            ))
+        }
+    }
+
+    // Similar to getitimer. Returns (next value, current value)
+    pub fn get_itimer(&self) -> (RustDuration, RustDuration) {
+        let guard = self._ac.lock().unwrap();
+
+        (guard.curr_duration, guard.next_duration)
+    }
+
+    fn _set_itimer(&self, guard: &mut MutexGuard<_IntervalTimer>, curr_duration: RustDuration, next_duration: RustDuration) {
+        if curr_duration.is_zero() {
+            guard.is_ticking = false;
+        } else {
+            guard.start_instant = RustInstant::now();
+            guard.curr_duration = curr_duration;
+            guard.next_duration = next_duration;
+
+            if !guard.is_ticking {
+                guard.is_ticking = true;
+
+                let self_dup = self.clone();
+                thread::spawn(move || { // There is a chance that there'll be two ticking threads running
+                                        // at the same time
+                    self_dup.tick();
+                });
+            }
+        }
+    }
+
+    pub fn set_itimer(&self, curr_duration: RustDuration, next_duration: RustDuration) {
+        let mut guard = self._ac.lock().unwrap();
+        self._set_itimer(&mut guard, curr_duration, next_duration);
+    }
+
+    pub fn tick(&self) {
+        loop {
```

Comment:
If I'm reading this right, once a ticking thread is launched it never terminates? Should we terminate the thread if it is cancelled.

Reply:
So inside the loop, it will check if is_ticking is true. If it's not then it breaks the loop and terminates the thread. The reason I don't make it a while loop is that it involves acquiring and releasing the mutex twice, which could cause some race conditions as is_ticking could be changed in between.